src/codegen/traverse.lisp

```diff
@@ -390,6 +443,27 @@ without any slot information."
         (format t "POST: ~v@{|   ~}~A~%" counter (class-name (class-of node)))
         (values))))))
 
+(defun print-node-parent (node)
+  "Print visiting node and its parent, using `traverse-with-parent-thunk`."
+  (declare (type node node)
+           (values node &optional))
+  (let ((counter 0))
+    (traverse-with-parent-thunk
+     node
+     (list
+      (action (:before node node parent-thunk)
+        (format t "PRE:  ~v@{|   ~}~A ~A~%" counter
+                (class-name (class-of node))
+                (class-name (class-of (funcall parent-thunk))))
+        (incf counter)
+        (values))
+      (action (:after node node parent-thunk)
+        (decf counter)
+        (format t "POST: ~v@{|   ~}~A ~A~%" counter
+                (class-name (class-of node))
+                (class-name (class-of (funcall parent-thunk))))
+        (values))))))
```

Comment:
As far as I can tell, identical behavior can be achieved without traversing with the extra hook:
(not actually a code suggestion, just formatting it as one for easy comparison)
```suggestion
(defun print-node-parent (node)
  "Print visiting node and its parent, using `traverse-with-parent-thunk`."
  (declare (type node node)
           (values node &optional))
  (let ((traversal-stack nil))
    (traverse
     node
     (list
      (action (:before node node)
        (format t "PRE:  ~v@{|   ~}~A ~A~%" (length traversal-stack)
                (class-name (class-of node))
                (class-name (class-of (car traversal-stack))))
        (push node traversal-stack)
        (values))
      (action (:after node node)
        (pop traversal-stack)
        (format t "POST: ~v@{|   ~}~A ~A~%" (length traversal-stack)
                (class-name (class-of node))
                (class-name (class-of (car traversal-stack))))
        (values))))))
```
Can you think of any other examples, where some functionality can only be achieved conveniently with `traverse-with-hook`, but not with the original `traverse`?

Reply:
You're completely right.   Merely obtaining parent can be achieved by this.
What I really need is a convenient API to access to the parent node, so we may avoid employing `*traverse-hook*` hack.  Let me think on it.